pymaker/auctions.py

```diff
@@ -660,3 +641,270 @@ def parse_event(self, event):
+
+    def cusp(self) -> Ray:
+        """Percentage drop before auction reset"""
+        return Ray(self._contract.functions.cusp().call())
+
+    def chip(self) -> Wad:
+        """Percentage of tab to suck from vow to incentivize keepers"""
+        return Wad(self._contract.functions.chip().call())
+
+    def tip(self) -> Rad:
+        """Flat fee to suck from vow to incentivize keepers"""
+        return Rad(self._contract.functions.tip().call())
+
+    def kicks(self) -> int:
+        """Number of auctions started so far."""
+        return int(self._contract.functions.kicks().call())
+
+    def active_count(self) -> int:
+        """Number of active and redoable auctions."""
+        return int(self._contract.functions.count().call())
+
+    def status(self, id: int) -> (bool, Ray, Wad, Rad):
+        """Indicates current state of the auction
+        Args:
+            id: Auction identifier.
+        """
+        assert isinstance(id, int)
+        (needs_redo, price, lot, tab) = self._contract.functions.getStatus(id).call()
+        logging.debug(f"Auction {id} {'needs redo ' if needs_redo else ''}with price={float(Ray(price))} " 
+                      f"lot={float(lot)} tab={float(tab)}")
+        return needs_redo, Ray(price), Wad(lot), Rad(tab)
+
+    def sales(self, id: int) -> Sale:
+        """Returns the auction details.
+        Args:
+            id: Auction identifier.
+        Returns:
+            The auction details.
+        """
+        assert(isinstance(id, int))
+
+        array = self._contract.functions.sales(id).call()
+
+        return Clipper.Sale(id=id,
+                            pos=int(array[0]),
+                            tab=Rad(array[1]),
+                            lot=Wad(array[2]),
+                            usr=Address(array[3]),
+                            tic=int(array[4]),
+                            top=Ray(array[5]))
+
+    def validate_take(self, id: int, amt: Wad, max: Ray, our_address: Address = None):
+        """Raise assertion if collateral cannot be purchased from an auction as desired"""
+        assert isinstance(id, int)
+        assert isinstance(amt, Wad)
+        assert isinstance(max, Ray)
+
+        if our_address:
+            assert isinstance(our_address, Address)
+        else:
+            our_address = Address(self.web3.eth.defaultAccount)
+
+        (done, price, lot, tab) = self.status(id)
+        assert not done
+        assert max >= price
+
+        slice: Wad = min(lot, amt)          # Purchase as much as possible, up to amt
+        owe: Rad = Rad(slice) * Rad(price)  # DAI needed to buy a slice of this sale
+
+        if Rad(owe) > tab:
+            owe = Rad(tab)
+            slice = Wad(owe / Rad(price))
+        elif owe < tab and slice < lot:
+            ilk = self.vat.ilk(self.ilk_name())
+            if (tab - owe) < ilk.dust:
+                assert tab > ilk.dust
+                owe = tab - ilk.dust
```

Comment:
We had a late-breaking change that uses the product of `chop` and `dust` instead of just `dust`, stored as a single quantity `chost` in the `Clipper`, that ended up going in for launch:
https://github.com/makerdao/dss/pull/222/files

You'll want to update this logic to read `Clipper.chost` and use it in place of `dust`.

Reply:
`dss-deploy-scripts` [deploys dss for liq-2.0](https://github.com/makerdao/dss-deploy-scripts/blob/master/.dapp.json#L280) at a commit which isn't producing an ABI with `clip.chost`.  As such, I'll continue testing as-is until `dss-deploy-scripts` produces the correct artifacts.

Reply:
I deployed a new testchain from https://github.com/makerdao/dss-deploy-scripts/tree/latest-upgrades-liq-2.0 and updated `pymaker` to wrap `chost`.  Feel free to re-review.